README.md
# CASSANDRA-LOADER

## Note on the fork

This forks adds features I needed to import csv data into cassandra.

Usage has changed, since I removed the custom build steps:

```
java -cp <jar file> com.datastax.loader.CqlDelimLoad -f <filename> -host <ipaddress> -schema <schema> [OPTIONS]
```
instead of
```
cassandra-loader -f <filename> -host <ipaddress> -schema <schema> [OPTIONS]
```

This fork also uses java code conventions and the gradle build doesn't rely on shell scripts and should run on any os.

## Introduction

cassandra-loader is a general-purpose, delimited-file, bulk loader for 
Cassandra. It supports a number of configuration options to enable bulk 
loading of various types of delimited files, including
* comma-separated values
* tab-separated values
* customer delimiter-separated values
* header row
* comma as decimal separator
* ...

## Getting it

### Downloading

This utility has already been built, and is available at
https://github.com/brianmhess/cassandra-loader/releases/download/v0.0.20/cassandra-loader

Get it with wget:

```
wget https://github.com/brianmhess/cassandra-loader/releases/download/v0.0.20/cassandra-loader
```

### Building

To build this repository, simply clone this repo and run:

```
gradle loader
```

All of the dependencies are included (namely, the Java driver - currently
version 3.0.0).  The output will be the cassandra-loader executable
in the build directory.  There will also be an jar with all of the
dependencies included in the build/libs/cassandra-loader-uber-<version>.jar

## Documentation 

To extract this README document, simply run (on the cassandra-loader
executable - (e.g., on build/cassandra-loader):

```
jar xf cassandra-loader README.md
```

## Run

To run cassandra-loader, simply run the cassandra-loader executable 
(e.g., located at build/cassandra-loader):

```
cassandra-loader
```

If you built this with gradle, you can also run:

```
gradle run
```

This will print the usage statement.

The following will load the `myFileToLoad.csv` file into the Cassandra 
cluster at IP address 1.2.3.4 into the `test.ltest` column family where 
the myFileToLoad file has the format of 4 columns - and it gets the
data type information from the database - and using the default options:

```
cassandra-loader -f myFileToLoad.csv -host 1.2.3.4 -schema "test.ltest(a, b, c, d)"
```

## Options:

 Switch           | Option             | Default                    | Description
-----------------:|-------------------:|---------------------------:|:----------
 `-configFile`    | Filename           | none                       | Filename of configuration options 
 `-f`             | Filename / Directory  | &lt;REQUIRED&gt;        | Filename or directory to load - required.
 `-filePattern`   | Filename Pattern          | none (all files)    | If directory is loaded this filters the files to load. The syntax is described in the [getPathMatcher JavaDoc](https://docs.oracle.com/javase/7/docs/api/java/nio/file/FileSystem.html#getPathMatcher(java.lang.String))
 `-host`          | IP Address         | &lt;REQUIRED&gt;           | Cassandra connection point - required.
 `-schema`        | CQL schema         | &lt;REQUIRED&gt;           | Schema of input data - required In the format "keySpace.table(col1,col2,...)" and in the order that the data will be in the file.
 `-port`          | Port Number        | 9042                       | Cassandra native protocol port number
 `-user`          | Username           | none                       | Cassandra username
 `-pw`            | Password           | none                       | Cassandra password
 `-ssl-truststore-path` | Truststore Path     | none                | Path to SSL truststore
 `-ssl-truststore-pwd`  | Truststore Password | none                | Password to SSL truststore
 `-ssl-keystore-path`   | Keystore Path       | none                | Path to SSL keystore
 `-ssl-keystore-path`   | Keystore Password   | none                | Password to SSL keystore
 `-consistencyLevel` | Consistency Level | ONE                       | CQL Consistency Level
 `-numThreads`    | Number of threads  | Number of CPUs             | Number of threads to use (one per file)
 `-numFutures`    | Number of Futures  | 1000                       | Number of Java driver futures in flight.
 `-numRetries`    | Number of retries  | 1                          | Number of times to retry the INSERT before declaring defeat.
 `-queryTimeout`  | Timeout in seconds | 2                          | Amount of time to wait for a query to finish before timing out.
 `-delim`         | Delimiter          | ,                          | Delimiter to use
 `-quote`         | Quote character          | "                          | Quote-character to use
 `-escape`         | Quote character          | \                          | Escape-character to use
 `-nullString`    | Null String        | &lt;empty string&gt;             | String to represent NULL data
 `-boolStyle`     | Boolean Style      | TRUE_FALSE                 | String for boolean values.  Options are "1_0", "Y_N", "T_F", "YES_NO", "TRUE_FALSE".
 `-decimalDelim`  | Decimal delimiter  | .                          | Delimiter for decimal values.  Options are "." or ","
 `-dateFormat`    | Date Format String | default for Locale.ENGLISH | Date format string as specified in the SimpleDateFormat Java class: http://docs.oracle.com/javase/7/docs/api/java/text/SimpleDateFormat.html
 `-skipRows`      | Rows to skip       | 0                          | Number of rows to skip at the beginning of the file
 `-skipCols`      | Columns to skip    | &lt;not set&gt;                  | Comma-separated list of columns to skip loading (0-counted)
 `-maxRows`       | Max rows to read   | -1                         | Maximum rows to read (after optional skipping of rows).  -1 signifies all rows.
 `-maxErrors`     | Max parse errors   | 10                         | Maximum number of rows that do not parse to allow before exiting.
 `-maxInsertErrors`| Max insert errors | 10                         | Maximum number of rows that do not insert to allow before exiting.
 `-badDir`        | Bad directory      | current directory          | Directory to write badly parsed and badly inserted rows - as well as the log file.
 `-rate`          | Ingest rate        | unlimited                  | Maximum rate to insert data - in rows/sec.
 `-progressRate`  | Progress rate      | 100000                     | How often to report the ingest rate (number of rows)
 `-rateFile`      | Rate Stats File    | &lt;not set&gt;                  | File to contain CSV rate statistics
 `-successDir`    | Success directory  | &lt;not set&gt;                  | Location to move successfully loaded files
 `-failureDir`    | Failure directory  | &lt;not set&gt;                  | Location to move files that failed to load
 `-batchSize`     | Batch size         | 1                          | Size of unlogged batches. If set to 1 then no batching.

## Comments

### Using stdin

You can send data in on stdin by specifying the filename (via the -f switch) as "stdin" (case insensitive).
That way, you could pipe data in from other commands:

```
grep IMPORTANT data.csv | cassandra-loader -f stdin -h 1.2.3.4 -schema "test.itest(a, b)"
```

### Support for collections 

Collections are supported.  Their format is the CQL native one.
Sets are started with '{' and ended with '}' and enclose a comma-separated list
{1,2,3} or {"a","b","c"}
Lists are started with '[' and ended with ']' and enclose a comma-separated list
[1,2,3] or ["a","b","c"]
Maps are started with '{' and ended with '}' and enclose a comma-separated list
of pairs that are separated by ':'
{1:1,2:2,3:3} or {"a":1, "b":2, "c":3}
All collections must be enclosed in double-quotes.

### Username/Password

If you specify either the username or the password, then you must specify both.

### Boolean Style

boolStyle is a case-insensitive test of the True and False strings.  For the
different styles, the True and False strings are as follows:

```
    Style   | True | False
------------|------|-------
     0_1    |    1 |     0 
     Y_N    |    Y |     N 
     T_F    |    T |     F 
   YES_NO   |  YES |    NO 
 TRUE_FALSE | TRUE | FALSE 
```

### Configuration file

configFile is a file with configuration options that are formatted just like on
the command line.  This allows you to not specify arguments on the command line.
For example, you can specify passwords in the configFile and avoid having them on
the command line.  The format is one switch and option per line:

```
-pwd mypassword
-host 1.2.3.4
```

### Miscelaneous

numFutures is a way to control the level of parallelism, but at some point 
too many will actually slow down the load.  The default of 500 is a decent 
place to start.

If you use quoted strings, you need to use double-quotes.
To escape a double-quote inside a quoted string, use the backslash to escape it ("\"") or specify the escape character using the `-escape` option.
To create the escape character inside a quoted string, use it twice.  If you quote your string, it will not be trimmed, but if you do not quote your string it will be trimmed.

#### Example
```
"Some people call me the ""space cowboy"", yeah"
```
with `-quote "` and `-escape "` results in
```
Some people call me the "space cowboy", yeah
```
Note that your shell might require additional quoting.

Loading into counter columns is not supported.

The default nullString is the empty string.  If you want empty strings to be saved as empty strings, set the nullString to something else.

If you do not set the successDir then files that successfully loaded will remain in their input directory.  The same is true for failed files if you do not set the failureDir.  You cannot set either if the input file is "stdin".


## Usage Statement:

```
version: 0.0.21-SNAPSHOT
Usage: -f <filename|directory> -host <ipaddress> -schema <schema> [OPTIONS]
OPTIONS:
  -configFile <filename>         File with configuration options
  -delim <delimiter>             Delimiter to use [,]
  -quote <quote>                 Quote character to use ["]
  -escape <escape>               Escape character to use [\]
  -dateFormat <dateFormatString> Date format [default for Locale.ENGLISH]
  -nullString <nullString>       String that signifies NULL [none]
  -skipRows <skipRows>           Number of rows to skip [0]
  -skipCols <columnsToSkip>      Comma-separated list of columsn to skip in the input file
  -maxRows <maxRows>             Maximum number of rows to read (-1 means all) [-1]
  -maxErrors <maxErrors>         Maximum parse errors to endure [10]
  -badDir <badDirectory>         Directory for where to place badly parsed rows. [none]
  -port <portNumber>             CQL Port Number [9042]
  -user <username>               Cassandra username [none]
  -pw <password>                 Password for user [none]
  -ssl-truststore-path <path>    Path to SSL truststore [none]
  -ssl-truststore-pwd <pwd>       Password for SSL truststore [none]
  -ssl-keystore-path <path>      Path to SSL keystore [none]
  -ssl-keystore-pwd <pwd>         Password for SSL keystore [none]
  -consistencyLevel <CL>         Consistency level [LOCAL_ONE]
  -numFutures <numFutures>       Number of CQL futures to keep in flight [1000]
  -batchSize <batchSize>         Number of INSERTs to batch together [1]
  -decimalDelim <decimalDelim>   Decimal delimiter [.] Other option is ','
  -boolStyle <boolStyleString>   Style for booleans [TRUE_FALSE]
  -numThreads <numThreads>       Number of concurrent threads (files) to load [num cores]
  -queryTimeout <# seconds>      Query timeout (in seconds) [2]
  -numRetries <numRetries>       Number of times to retry the INSERT [1]
  -maxInsertErrors <# errors>    Maximum INSERT errors to endure [10]
  -rate <rows-per-second>        Maximum insert rate [50000]
  -progressRate <num txns>       How often to report the insert rate [100000]
  -rateFile <filename>           Where to print the rate statistics
  -successDir <dir>              Directory where to move successfully loaded files
  -failureDir <dir>              Directory where to move files that did not successfully load
  -nullsUnset [false|true]       Treat nulls as unset [false]
  -maxCharsPerColumn <int>       Buffer size for parsing columns [4096]
  -filePattern <pattern>         When -f is a folder: use only files matching this pattern [all files]



Examples:
cassandra-loader -f /path/to/file.csv -host localhost -schema "test.test3(a, b, c)"
cassandra-loader -f /path/to/directory -host 1.2.3.4 -schema "test.test3(a, b, c)" -delim "\t" -numThreads 10
cassandra-loader -f stdin -host localhost -schema "test.test3(a, b, c)" -user myuser -pw mypassword
```

##Examples:

Load file /path/to/file.csv into the test3 table in the test keyspace using
the cluster at localhost.  Use the default options:

```
cassandra-loader -f /path/to/file.csv -host localhost -schema "test.test3(a, b, c)"
```

Load all the files from /path/to/directory into the test3 table in the test
keyspace using the cluster at 1.2.3.4.  Use 10 threads and use tab as the
delimiter:

```
cassandra-loader -f /path/to/directory -host 1.2.3.4 -schema "test.test3(a, b, c)" -delim "\t" -numThreads 10
```

Load the data from stdin into the test3 table in the test keyspace using the
cluster at localhost.  Use "myuser" as the username and "mypassword" as the
password:

```
cassandra-loader -f stdin -host localhost -schema "test.test3(a, b, c)" -user myuser -pw mypassword
```

##Sample

Included here is a set of sample data.  It is in the sample/ directory.
You can set up the table and keyspace by running:

```
cqlsh -f sample/cassandra-schema.cql
```

To load the data, run:

```
cd sample
./load.sh
```

To check that things have succeeded, you can run:

```
wc -l titanic.csv
```

And:

```
cqlsh -e "SELECT COUNT(*) FROM titanic.surviors"
```

Both should return 891.



## cassandra-unloader
cassandra-unloader is a utility to dump the contents
of a Cassandra table to delimited file format.  It uses
the same sorts of options as cassandra-loader so that the
output of cassandra-unloader could be piped into 
cassandra-loader:

```
cassandra-unloader -f stdout -host host1 -schema "ks.table(a,b,c)" | cassandra-loader -f stdin -host host2 -schema "ks2.table2(x,y,z)"
```

Get it with wget:
```
wget https://github.com/brianmhess/cassandra-loader/releases/download/v0.0.20/cassandra-unloader
```

To build, run:

```
gradle unloader
```

To run cassandra-unloader, simply run the cassandra-unloader executable 
(e.g., located at build/cassandra-unloader):

```
cassandra-unloader
```

###Usage statement:

```
version: 0.0.20
Usage: -f <outputStem> -host <ipaddress> -schema <schema> [OPTIONS]
OPTIONS:
  -configFile <filename>         File with configuration options
  -delim <delimiter>             Delimiter to use [,]
  -dateFormat <dateFormatString> Date format [default for Locale.ENGLISH]
  -nullString <nullString>       String that signifies NULL [none]
  -port <portNumber>             CQL Port Number [9042]
  -user <username>               Cassandra username [none]
  -pw <password>                 Password for user [none]
  -ssl-truststore-path <path>    Path to SSL truststore [none]
  -ssl-truststore-pwd <pwd>       Password for SSL truststore [none]
  -ssl-keystore-path <path>      Path to SSL keystore [none]
  -ssl-keystore-pwd <pwd>         Password for SSL keystore [none]
  -consistencyLevel <CL>         Consistency level [LOCAL_ONE]
  -decimalDelim <decimalDelim>   Decimal delimiter [.] Other option is ','
  -boolStyle <boolStyleString>   Style for booleans [TRUE_FALSE]
  -numThreads <numThreads>       Number of concurrent threads to unload [5]
  -beginToken <tokenString>      Begin token [none]
  -endToken <tokenString>        End token [none]
  -where <predicate>             WHERE clause [none]
```

A few simple examples using the `-where` are as follows:

```
cassandra-unloader -host localhost -f stdout -schema "testks.testtable(pkey,ccol,x,y)" -where "pkey=5"
cassandra-unloader -host localhost -f stdout -schema "testks.testtable(pkey,ccol,x,y)" -where "x = 100 ALLOW FILTERING"
```


ChangeLog.md
## 0.0.21-SNAPSHOT
- Added -quote
- Added -escape
- Added -maxCharsPerColumn
- Added -filePattern

## 0.0.20
- Fixed delimiter in MapParser
- Catch NULLs in Map/Set/List parsing and throw in BADPARSE

## 0.0.19
- Converted to parse with Univocity CSV parser
- Added -where

## 0.0.18
- Support for Cassandra 3.0

## 0.0.17
- Fixed null collection issue / NPE (Issue 8)

## 0.0.16
- Unloader will quote collections (which the loader expects)
- Fixed collection issue (Issue 14)
- BLOBs are now Base64 encoded on unload, and should be Base64 to load (Issue 15)
- Support for quoted keyspace, table, and column names

## 0.0.15
- Better error handling for case when C* inserts are failing

## 0.0.14
- Updated cassandra-unloader to add support for collections,
	consistency level, ssl, etc

## 0.0.13
- Added configFile
- added ssl options (with truststore and keystore)

## 0.0.12
- Added a rateFile to output CSV rate statistics
- added -skipCols to skip input columns

## 0.0.11
- Added support for quoted Keyspaces, Tables, and Columns

## 0.0.10
- You want collections?  You got 'em
- Added progress reporting - you can specify the rate at which
	the rate is reported via the -progressRate option
- Refactored RateLimiting - added it to a new RateLimitingSession
- Laid groundwork for Dynamic rate limiting - to be worked out
	once we find a way to collect the right statistic

## 0.0.9
- Added -successDir and -failureDir
- Added return codes for the loader and unloader
- Refactored BoolStyle
- Cleaned up the readme a bit



src/main/java/com/datastax/loader/CqlDelimLoad.java
/*
 * Copyright 2015 Brian Hess
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.datastax.loader;

import com.datastax.driver.core.Cluster;
import com.datastax.driver.core.Metrics;
import com.datastax.driver.core.ProtocolVersion;
import com.datastax.driver.core.Session;

import java.io.*;
import java.nio.file.FileSystems;
import java.nio.file.PathMatcher;
import java.security.KeyManagementException;
import java.security.KeyStoreException;
import java.security.NoSuchAlgorithmException;
import java.security.UnrecoverableKeyException;
import java.security.cert.CertificateException;
import java.text.ParseException;
import java.util.*;
import java.util.concurrent.*;

public class CqlDelimLoad extends ConfigurationLoader {
    public static final String STDIN = "stdin";
    public static final String STDERR = "stderr";
    private int numFutures = 1000;
    private int inNumFutures = -1;
    private int queryTimeout = 2;
    private long maxInsertErrors = 10;
    private int numRetries = 1;
    private double rate = 50000.0;
    private long progressRate = 100000;
    private RateLimiter rateLimiter = null;
    private String rateFile = null;
    private PrintStream rateStream = null;
    private long maxErrors = 10;
    private long skipRows = 0;
    private String skipCols = null;
    private long maxRows = -1;
    private String badDir = ".";
    private String successDir = null;
    private String failureDir = null;

    private String filePattern = null;

    private int numThreads = Runtime.getRuntime().availableProcessors();
    private int batchSize = 1;
    private boolean nullsUnset = false;

    public static void main(String[] args)
            throws IOException, ParseException, InterruptedException, ExecutionException,
            KeyStoreException, NoSuchAlgorithmException, UnrecoverableKeyException,
            CertificateException, KeyManagementException {
        CqlDelimLoad cdl = new CqlDelimLoad();
        boolean success = cdl.run(args);
        if (success) {
            System.exit(0);
        } else {
            System.exit(-1);
        }
    }


    private String usage() {
        StringBuilder usage = new StringBuilder("version: ").append(version).append("\n");
        usage.append("Usage: -f <filename|directory> -host <ipaddress> -schema <schema> [OPTIONS]\n");
        usage.append("OPTIONS:\n");
        usage.append(commonUsage());

        usage.append("  -skipRows <skipRows>           Number of rows to skip [0]\n");
        usage.append("  -skipCols <columnsToSkip>      Comma-separated list of columsn to skip in the input file\n");
        usage.append("  -maxRows <maxRows>             Maximum number of rows to read (-1 means all) [-1]\n");
        usage.append("  -maxErrors <maxErrors>         Maximum parse errors to endure [10]\n");
        usage.append("  -badDir <badDirectory>         Directory for where to place badly parsed rows. [none]\n");
        usage.append("  -numFutures <numFutures>       Number of CQL futures to keep in flight [1000]\n");
        usage.append("  -batchSize <batchSize>         Number of INSERTs to batch together [1]\n");
        usage.append("  -queryTimeout <# seconds>      Query timeout (in seconds) [2]\n");
        usage.append("  -numRetries <numRetries>       Number of times to retry the INSERT [1]\n");
        usage.append("  -maxInsertErrors <# errors>    Maximum INSERT errors to endure [10]\n");
        usage.append("  -rate <rows-per-second>        Maximum insert rate [50000]\n");
        usage.append("  -progressRate <num txns>       How often to report the insert rate [100000]\n");
        usage.append("  -rateFile <filename>           Where to print the rate statistics\n");
        usage.append("  -successDir <dir>              Directory where to move successfully loaded files\n");
        usage.append("  -failureDir <dir>              Directory where to move files that did not successfully load\n");
        usage.append("  -nullsUnset [false|true]       Treat nulls as unset [false]\n");
        usage.append("  -filePattern <pattern>         When -f is a folder: use only files matching this pattern [all files]\n");

        usage.append("\n\nExamples:\n");
        usage.append("cassandra-loader -f /path/to/file.csv -host localhost -schema \"test.test3(a, b, c)\"\n");
        usage.append("cassandra-loader -f /path/to/directory -host 1.2.3.4 -schema \"test.test3(a, b, c)\" -delim \"\\t\" -numThreads 10\n");
        usage.append("cassandra-loader -f stdin -host localhost -schema \"test.test3(a, b, c)\" -user myuser -pw mypassword\n");
        return usage.toString();
    }

    @Override
    protected boolean validateArgs() {
        super.validateArgs();
        if (0 >= numFutures) {
            System.err.println("Number of futures must be positive (" + numFutures + ")");
            return false;
        }
        if (0 >= batchSize) {
            System.err.println("Batch size must be positive (" + batchSize + ")");
            return false;
        }
        if (0 >= queryTimeout) {
            System.err.println("Query timeout must be positive");
            return false;
        }
        if (0 > maxInsertErrors) {
            System.err.println("Maximum number of insert errors must be non-negative");
            return false;
        }
        if (0 > numRetries) {
            System.err.println("Number of retries must be non-negative");
            return false;
        }
        if (0 > skipRows) {
            System.err.println("Number of rows to skip must be non-negative");
            return false;
        }
        if (0 >= maxRows) {
            System.err.println("Maximum number of rows to load must be positive");
            return false;
        }
        if (0 > maxErrors) {
            System.err.println("Maximum number of parse errors must be non-negative");
            return false;
        }
        if (0 > progressRate) {
            System.err.println("Progress rate must be non-negative");
            return false;
        }
        if (!STDIN.equalsIgnoreCase(filename)) {
            File infile = new File(filename);
            if ((!infile.isFile()) && (!infile.isDirectory())) {
                System.err.println("The -f argument needs to be a file or a directory");
                return false;
            }
            if (infile.isDirectory()) {
                File[] infileList = infile.listFiles();
                if (infileList.length < 1) {
                    System.err.println("The directory supplied is empty");
                    return false;
                }
            }
        }
        if (null != successDir) {
            if (STDIN.equalsIgnoreCase(filename)) {
                System.err.println("Cannot specify -successDir with stdin");
                return false;
            }
            File sdir = new File(successDir);
            if (!sdir.isDirectory()) {
                System.err.println("-successDir must be a directory");
                return false;
            }
        }
        if (null != failureDir) {
            if (STDIN.equalsIgnoreCase(filename)) {
                System.err.println("Cannot specify -failureDir with stdin");
                return false;
            }
            File sdir = new File(failureDir);
            if (!sdir.isDirectory()) {
                System.err.println("-failureDir must be a directory");
                return false;
            }
        }

        if (0 > rate) {
            System.err.println("Rate must be positive");
            return false;
        }

        return true;
    }

    protected boolean parseArgsFromMap(Map<String, String> amap) {
        super.parseArgsFromMap(amap);

        String tkey;
        if (null != (tkey = amap.remove("-numFutures"))) inNumFutures = Integer.parseInt(tkey);
        if (null != (tkey = amap.remove("-batchSize"))) batchSize = Integer.parseInt(tkey);
        if (null != (tkey = amap.remove("-queryTimeout"))) queryTimeout = Integer.parseInt(tkey);
        if (null != (tkey = amap.remove("-maxInsertErrors"))) maxInsertErrors = Long.parseLong(tkey);
        if (null != (tkey = amap.remove("-numRetries"))) numRetries = Integer.parseInt(tkey);
        if (null != (tkey = amap.remove("-maxErrors"))) maxErrors = Long.parseLong(tkey);
        if (null != (tkey = amap.remove("-skipRows"))) skipRows = Integer.parseInt(tkey);
        if (null != (tkey = amap.remove("-skipCols"))) skipCols = tkey;
        if (null != (tkey = amap.remove("-maxRows"))) maxRows = Integer.parseInt(tkey);
        if (null != (tkey = amap.remove("-badDir"))) badDir = tkey;
        if (null != (tkey = amap.remove("-filePattern"))) {
            try {
                FileSystems.getDefault().getPathMatcher(tkey);
            } catch (Throwable ignored) {
                System.err.println("Bad filePattern parameter. See https://docs.oracle.com/javase/7/docs/api/java/nio/file/FileSystem.html#getPathMatcher(java.lang.String) for usage");
                return false;
            }
            filePattern = tkey;
        }
        if (null != (tkey = amap.remove("-numThreads"))) numThreads = Integer.parseInt(tkey);
        if (null != (tkey = amap.remove("-rate"))) rate = Double.parseDouble(tkey);
        if (null != (tkey = amap.remove("-progressRate"))) progressRate = Long.parseLong(tkey);
        if (null != (tkey = amap.remove("-rateFile"))) rateFile = tkey;
        if (null != (tkey = amap.remove("-successDir"))) successDir = tkey;
        if (null != (tkey = amap.remove("-failureDir"))) failureDir = tkey;
        if (null != (tkey = amap.remove("-nullsUnset"))) nullsUnset = Boolean.parseBoolean(tkey);

        if (-1 == maxRows) {
            maxRows = Long.MAX_VALUE;
        }
        if (-1 == maxErrors) {
            maxErrors = Long.MAX_VALUE;
        }
        if (-1 == maxInsertErrors) {
            maxInsertErrors = Long.MAX_VALUE;
        }

        if (!amap.isEmpty()) {
            for (String k : amap.keySet()) {
                System.err.println("Unrecognized option: " + k);
            }
            return false;
        }

        if (0 < inNumFutures) {
            numFutures = inNumFutures / numThreads;
        }

        return true;
    }

    @Override
    protected int getNumConnections() {
        return 8;
    }

    @Override
    protected Session getSession(Cluster cluster) throws FileNotFoundException {
        Session tsession = cluster.connect();

        if ((0 > cluster.getConfiguration().getProtocolOptions()
                .getProtocolVersion().compareTo(ProtocolVersion.V4))
                && nullsUnset) {
            System.err.println("Cannot use nullsUnset with ProtocolVersion less than V4 (prior to Cassandra 3.0");
            cleanup();
            return null;
        }

        if (null != rateFile) {
            if (STDERR.equalsIgnoreCase(rateFile)) {
                rateStream = System.err;
            } else {
                rateStream = new PrintStream(new BufferedOutputStream(new FileOutputStream(rateFile)), true);
            }
        }
        Metrics metrics = cluster.getMetrics();
        com.codahale.metrics.Timer timer = metrics.getRequestsTimer();
        rateLimiter = new RateLimiter(rate, progressRate, timer, rateStream);
        //rateLimiter = new Latency999RateLimiter(rate, progressRate, 3000, 200, 10, 0.5, 0.1, cluster, false);
        session = new RateLimitedSession(tsession, rateLimiter);
        return session;
    }

    @Override
    protected void cleanup() {
        super.cleanup();
        rateLimiter.report(null, null);
        if (null != rateStream)
            rateStream.close();
    }

    public boolean run(String[] args)
            throws IOException, ParseException, InterruptedException, ExecutionException, KeyStoreException,
            NoSuchAlgorithmException, KeyManagementException, CertificateException,
            UnrecoverableKeyException {
        if (!parseArgs(args)) {
            System.err.println("Bad arguments");
            System.err.println(usage());
            return false;
        }

        // Setup
        if (!setup())
            return false;

        // open file
        Deque<File> fileList = new ArrayDeque<>();
        File infile;
        File[] inFileList;
        boolean onefile = true;
        if (STDIN.equalsIgnoreCase(filename)) {
            infile = null;
        } else {
            infile = new File(filename);
            if (!infile.isFile()) {
                inFileList = infile.listFiles();
                if (inFileList.length < 1)
                    throw new IOException("directory is empty");
                onefile = false;
                Arrays.sort(inFileList,
                        (f1, f2) -> f1.getName().compareTo(f2.getName()));
                PathMatcher matcher = null;
                if (filePattern != null) {
                    matcher = FileSystems.getDefault().getPathMatcher(filePattern);
                }
                for (final File file : inFileList) {
                    if (matcher == null || matcher.matches(file.toPath().getFileName())) {
                        fileList.push(file);
                    }
                }
            }
        }

        final CqlDelimParser cqlDelimParser = new CqlDelimParser(cqlSchema, delimiter, nullString,
                dateFormatString, boolStyle, locale,
                skipCols, session, true, quote, escape, maxCharsPerColumn);

        // Launch Threads
        ExecutorService executor;
        long total = 0;
        if (onefile) {
            // One file/stdin to process
            executor = Executors.newSingleThreadExecutor();
            Callable<Long> worker = getWorker(infile, cqlDelimParser);
            Future<Long> res = executor.submit(worker);
            total = res.get();
            executor.shutdown();
        } else {
            executor = Executors.newFixedThreadPool(numThreads);
            Set<Future<Long>> results = new HashSet<>();
            while (!fileList.isEmpty()) {
                File tFile = fileList.pop();
                Callable<Long> worker = getWorker(tFile, cqlDelimParser);
                results.add(executor.submit(worker));
            }
            executor.shutdown();
            for (Future<Long> res : results)
                total += res.get();
        }

        // Cleanup
        cleanup();
        //System.err.println("Total rows inserted: " + total);

        return true;
    }

    private Callable<Long> getWorker(File tFile, CqlDelimParser cqlDelimParser) {
        return new CqlDelimLoadTask(
                maxErrors, skipRows,
                maxRows, badDir, tFile,
                session,
                consistencyLevel,
                numFutures, batchSize,
                numRetries,
                queryTimeout,
                maxInsertErrors,
                successDir, failureDir,
                nullsUnset, cqlDelimParser);
    }
}

